File: modules/traffic-api/src/backend/services/traffic_generator.py
import asyncio
import random
import time
from datetime import datetime
from typing import Dict, List, Optional, Any
from uuid import uuid4
import math

class TrafficGenerator:
    """Traffic generation service that simulates dynamic traffic patterns on a graph network."""
    
    def __init__(self, config: Optional[Dict[str, Any]] = None):
        self.config = {
            'node_count': config.get('node_count', 20) if config else 20,
            'update_interval': config.get('update_interval', 30000) if config else 30000,  # ms
            'base_traffic_level': config.get('base_traffic_level', 0.15) if config else 0.15,
            'traffic_variability': config.get('traffic_variability', 0.35) if config else 0.35,
            'incident_probability': config.get('incident_probability', 0.00001) if config else 0.00001,
            'road_density': config.get('road_density', 0.3) if config else 0.3
        }
        
        self.vertices: List[Dict[str, Any]] = []
        self.edges: List[Dict[str, Any]] = []
        self.incidents: List[Dict[str, Any]] = []
        self.update_task: Optional[asyncio.Task] = None
        self.start_time = time.time()
        self.update_count = 0
        self.callbacks: List[callable] = []
        
        self._initialize_graph()
    
    def _initialize_graph(self):
        """Initialize the graph structure with vertices and edges."""
        # Create vertices (intersections/nodes)
        self.vertices = []
        for i in range(self.config['node_count']):
            self.vertices.append({
                'id': f'node_{i}',
                'name': f'Intersection {i + 1}',
                'x': random.random() * 800 + 50,
                'y': random.random() * 500 + 50
            })
        
        # Create edges (roads between nodes)
        self.edges = []
        edge_set = set()
        
        # Connect each node to a few random nearby nodes
        for i in range(len(self.vertices)):
            node_a = self.vertices[i]
            
            # Calculate distances to other nodes
            distances = []
            for j, node_b in enumerate(self.vertices):
                if i == j:
                    distances.append({'index': j, 'distance': float('inf')})
                else:
                    dx = node_a['x'] - node_b['x']
                    dy = node_a['y'] - node_b['y']
                    distances.append({'index': j, 'distance': math.sqrt(dx * dx + dy * dy)})
            
            # Sort by distance
            distances.sort(key=lambda x: x['distance'])
            
            # Connect to 2-4 nearest neighbors
            connect_count = 2 + random.randint(0, 2)
            for k in range(min(connect_count, len(distances))):
                j = distances[k]['index']
                edge_id = '-'.join(map(str, sorted([i, j])))
                
                if edge_id not in edge_set and random.random() < (self.config['road_density'] + 0.5):
                    edge_set.add(edge_id)
                    
                    distance = distances[k]['distance']
                    traffic = self._generate_random_traffic()
                    self.edges.append({
                        'id': f'edge_{len(self.edges)}',
                        'from': node_a['id'],
                        'to': self.vertices[j]['id'],
                        'weight': traffic,
                        'distance': round(distance / 10),
                        'speed': self._calculate_speed(traffic),
                        'vehicleCount': random.randint(0, 49),
                        'isIncident': False,
                        'incidentSeverity': 0,
                        'bidirectional': True
                    })
        
        # Ensure graph connectivity
        self._ensure_connectivity()
    
    def _ensure_connectivity(self):
        """Ensure all nodes are connected to at least one other node."""
        connected_nodes = set()
        
        for edge in self.edges:
            connected_nodes.add(edge['from'])
            connected_nodes.add(edge['to'])
        
        # Connect isolated nodes
        for vertex in self.vertices:
            if vertex['id'] not in connected_nodes:
                # Find nearest connected node
                nearest = None
                min_dist = float('inf')
                
                for other in self.vertices:
                    if other['id'] in connected_nodes:
                        dx = vertex['x'] - other['x']
                        dy = vertex['y'] - other['y']
                        dist = math.sqrt(dx * dx + dy * dy)
                        if dist < min_dist:
                            min_dist = dist
                            nearest = other
                
                if nearest:
                    traffic = self._generate_random_traffic()
                    self.edges.append({
                        'id': f'edge_{len(self.edges)}',
                        'from': vertex['id'],
                        'to': nearest['id'],
                        'weight': traffic,
                        'distance': round(min_dist / 10),
                        'speed': self._calculate_speed(traffic),
                        'vehicleCount': random.randint(0, 49),
                        'isIncident': False,
                        'incidentSeverity': 0,
                        'bidirectional': True
                    })
                    connected_nodes.add(vertex['id'])
    
    def _generate_random_traffic(self) -> float:
        """Generate random traffic level."""
        base = self.config['base_traffic_level']
        variability = self.config['traffic_variability']
        random_factor = (random.random() - 0.5) * variability * 4
        return max(0.0, min(1.0, base + random_factor))
    
    def _calculate_speed(self, traffic_level: float) -> int:
        """Calculate speed based on traffic level."""
        return round(60 - (traffic_level * 55))
    
    async def _update_traffic(self):
        """Update traffic state for all edges."""
        self.update_count += 1
        
        # Time-based patterns (rush hour simulation)
        hour = datetime.now().hour
        time_multiplier = 1.0
        
        if 7 <= hour <= 9:  # Morning rush
            time_multiplier = 1.5
        elif 17 <= hour <= 19:  # Evening rush
            time_multiplier = 1.6
        elif hour >= 23 or hour <= 5:  # Night time
            time_multiplier = 0.4
        
        # Update each edge
        for edge in self.edges:
            avg_neighbor_traffic = self._get_average_neighbor_traffic(edge)
            
            random_change = (random.random() - 0.5) * 0.2
            new_traffic = edge['weight'] * 0.85 + avg_neighbor_traffic * 0.05 + random_change
            
            # Occasional minor adjustments
            if random.random() < 0.05:
                new_traffic += (random.random() - 0.5) * 0.15
            
            # Apply time multiplier
            new_traffic *= time_multiplier
            
            # Clamp between 0 and 1
            new_traffic = max(0.0, min(1.0, new_traffic))
            
            edge['weight'] = new_traffic
            edge['speed'] = self._calculate_speed(new_traffic)
            
            # Update vehicle count
            base_vehicles = int(new_traffic * 100)
            edge['vehicleCount'] = max(0, base_vehicles + random.randint(-5, 5))
            
            # Random incident generation
            if not edge['isIncident'] and random.random() < self.config['incident_probability']:
                self.create_incident(
                    edge['id'],
                    random.random() * 0.5 + 0.3,
                    30000 + random.random() * 60000
                )
        
        # Update incidents
        self._update_incidents()
        
        # Notify callbacks
        await self._emit_update()
    
    def _get_average_neighbor_traffic(self, edge: Dict[str, Any]) -> float:
        """Get average traffic of connected edges."""
        connected_edges = [
            e for e in self.edges
            if e['id'] != edge['id'] and (
                e['from'] == edge['from'] or e['from'] == edge['to'] or
                e['to'] == edge['from'] or e['to'] == edge['to']
            )
        ]
        
        if not connected_edges:
            return self.config['base_traffic_level']
        
        total_traffic = sum(e['weight'] for e in connected_edges)
        return total_traffic / len(connected_edges)
    
    def create_incident(
        self,
        edge_id: str,
        severity: float = 0.5,
        duration: float = 60000,
        incident_type: Optional[str] = None
    ) -> Dict[str, Any]:
        """Create a traffic incident on an edge."""
        edge = next((e for e in self.edges if e['id'] == edge_id), None)
        
        if not edge:
            raise ValueError(f'Edge not found: {edge_id}')
        
        if edge['isIncident']:
            raise ValueError(f'Edge {edge_id} already has an active incident')
        
        # Validate severity and duration
        validated_severity = max(0.0, min(1.0, severity))
        validated_duration = max(10000, min(3600000, duration))
        
        incident = {
            'id': f'incident_{int(time.time() * 1000)}_{uuid4().hex[:9]}',
            'edgeId': edge_id,
            'from': edge['from'],
            'to': edge['to'],
            'severity': validated_severity,
            'startTime': int(time.time() * 1000),
            'endTime': int(time.time() * 1000) + int(validated_duration),
            'type': incident_type or self._get_random_incident_type(),
            'description': self._get_incident_description(
                incident_type or self._get_random_incident_type(),
                validated_severity
            )
        }
        
        self.incidents.append(incident)
        
        edge['isIncident'] = True
        edge['incidentSeverity'] = incident['severity']
        
        # Apply incident effect
        self._apply_incident_effect(edge_id, incident['severity'])
        
        return incident
    
    def _get_random_incident_type(self) -> str:
        """Get a random incident type."""
        types = ['accident', 'construction', 'roadblock', 'weather', 'event']
        return random.choice(types)
    
    def _get_incident_description(self, incident_type: str, severity: float) -> str:
        """Get incident description based on type and severity."""
        severity_text = 'severe' if severity > 0.7 else 'moderate' if severity > 0.4 else 'minor'
        severity_cap = severity_text.capitalize()
        
        descriptions = {
            'accident': f'{severity_cap} vehicle accident causing delays',
            'construction': f'{severity_cap} road construction work',
            'roadblock': f'{severity_cap} road closure or blockage',
            'weather': f'{severity_cap} weather-related conditions',
            'event': f'{severity_cap} special event causing congestion'
        }
        return descriptions.get(incident_type, f'{severity_cap} traffic incident')
    
    def _apply_incident_effect(self, edge_id: str, severity: float):
        """Apply traffic increase due to incident."""
        edge = next((e for e in self.edges if e['id'] == edge_id), None)
        if not edge:
            return
        
        # Increase traffic on incident road
        edge['weight'] = min(1.0, edge['weight'] + severity * 0.8)
        
        # Increase traffic on connected roads
        connected_edges = [
            e for e in self.edges
            if e['id'] != edge['id'] and (
                e['from'] == edge['from'] or e['from'] == edge['to'] or
                e['to'] == edge['from'] or e['to'] == edge['to']
            )
        ]
        
        for connected_edge in connected_edges:
            connected_edge['weight'] = min(1.0, connected_edge['weight'] + severity * 0.5)
    
    def _update_incidents(self):
        """Update and clear expired incidents."""
        now = int(time.time() * 1000)
        incidents_to_remove = []
        
        for incident in self.incidents:
            if now >= incident['endTime']:
                edge = next((e for e in self.edges if e['id'] == incident['edgeId']), None)
                if edge:
                    edge['isIncident'] = False
                    edge['incidentSeverity'] = 0
                incidents_to_remove.append(incident)
        
        for incident in incidents_to_remove:
            self.incidents.remove(incident)
    
    def clear_incident(self, incident_id: str) -> Dict[str, Any]:
        """Manually clear a specific incident."""
        incident = next((i for i in self.incidents if i['id'] == incident_id), None)
        
        if not incident:
            raise ValueError(f'Incident not found: {incident_id}')
        
        edge = next((e for e in self.edges if e['id'] == incident['edgeId']), None)
        if edge:
            edge['isIncident'] = False
            edge['incidentSeverity'] = 0
        
        self.incidents.remove(incident)
        return incident
    
    def get_current_state(self) -> Dict[str, Any]:
        """Get current state of the traffic system."""
        return {
            'vertices': self.vertices,
            'edges': [
                {
                    'id': e['id'],
                    'from': e['from'],
                    'to': e['to'],
                    'weight': e['weight'],
                    'distance': e['distance'],
                    'speed': e['speed'],
                    'vehicleCount': e['vehicleCount'],
                    'isIncident': e['isIncident'],
                    'incidentSeverity': e['incidentSeverity'],
                    'bidirectional': e['bidirectional']
                }
                for e in self.edges
            ],
            'incidents': self.incidents,
            'config': self.config,
            'updateCount': self.update_count,
            'timestamp': int(time.time() * 1000)
        }
    
    def get_edge_data(self, edge_id: str) -> Optional[Dict[str, Any]]:
        """Get data for a specific edge."""
        return next((e for e in self.edges if e['id'] == edge_id), None)
    
    def get_vertex_data(self, vertex_id: str) -> Optional[Dict[str, Any]]:
        """Get data for a specific vertex."""
        return next((v for v in self.vertices if v['id'] == vertex_id), None)
    
    def get_all_edges(self) -> List[Dict[str, Any]]:
        """Get all edges with basic info."""
        return [
            {
                'id': e['id'],
                'from': e['from'],
                'to': e['to'],
                'distance': e['distance'],
                'currentWeight': e['weight'],
                'hasIncident': e['isIncident']
            }
            for e in self.edges
        ]
    
    def get_available_edges_for_incident(self) -> List[Dict[str, Any]]:
        """Get edges available for incident creation."""
        return [
            {
                'id': e['id'],
                'from': e['from'],
                'to': e['to'],
                'distance': e['distance'],
                'currentTraffic': e['weight']
            }
            for e in self.edges if not e['isIncident']
        ]
    
    def get_active_incidents(self) -> List[Dict[str, Any]]:
        """Get all active incidents."""
        now = int(time.time() * 1000)
        return [
            {
                **i,
                'remainingTime': max(0, i['endTime'] - now)
            }
            for i in self.incidents
        ]
    
    def get_config(self) -> Dict[str, Any]:
        """Get current configuration."""
        return self.config.copy()
    
    def update_config(self, new_config: Dict[str, Any]):
        """Update configuration."""
        needs_reinit = False
        
        if 'node_count' in new_config and new_config['node_count'] != self.config['node_count']:
            self.config['node_count'] = new_config['node_count']
            needs_reinit = True
        
        if 'road_density' in new_config:
            self.config['road_density'] = new_config['road_density']
            needs_reinit = True
        
        if needs_reinit:
            self._initialize_graph()
        
        if 'update_interval' in new_config:
            self.config['update_interval'] = new_config['update_interval']
        
        if 'base_traffic_level' in new_config:
            self.config['base_traffic_level'] = new_config['base_traffic_level']
        
        if 'traffic_variability' in new_config:
            self.config['traffic_variability'] = new_config['traffic_variability']
        
        if 'incident_probability' in new_config:
            self.config['incident_probability'] = new_config['incident_probability']
    
    def get_stats(self) -> Dict[str, Any]:
        """Get traffic statistics."""
        total_edges = len(self.edges)
        total_vertices = len(self.vertices)
        total_traffic = sum(e['weight'] for e in self.edges)
        total_vehicles = sum(e['vehicleCount'] for e in self.edges)
        avg_speed = sum(e['speed'] for e in self.edges)
        
        return {
            'totalVertices': total_vertices,
            'totalEdges': total_edges,
            'averageTrafficLevel': f"{(total_traffic / total_edges):.3f}" if total_edges > 0 else "0.000",
            'totalVehicles': total_vehicles,
            'averageSpeed': round(avg_speed / total_edges) if total_edges > 0 else 0,
            'activeIncidents': len(self.incidents),
            'updateCount': self.update_count,
            'uptime': int((time.time() - self.start_time) * 1000)
        }
    
    def on_update(self, callback: callable):
        """Register a callback for traffic updates."""
        self.callbacks.append(callback)
    
    async def _emit_update(self):
        """Emit update to all registered callbacks."""
        state = self.get_current_state()
        for callback in self.callbacks:
            try:
                if asyncio.iscoroutinefunction(callback):
                    await callback(state)
                else:
                    callback(state)
            except Exception as e:
                print(f"Error in update callback: {e}")
    
    async def start(self):
        """Start the traffic generator."""
        if self.update_task and not self.update_task.done():
            print('Traffic generator already running')
            return
        
        print('Starting traffic generator...')
        
        async def update_loop():
            while True:
                await self._update_traffic()
                await asyncio.sleep(self.config['update_interval'] / 1000)
        
        self.update_task = asyncio.create_task(update_loop())
    
    async def stop(self):
        """Stop the traffic generator."""
        if self.update_task and not self.update_task.done():
            self.update_task.cancel()
            try:
                await self.update_task
            except asyncio.CancelledError:
                pass
            self.update_task = None
            print('Traffic generator stopped')

File: trafficGenerator.js
const EventEmitter = require('events');

class TrafficGenerator extends EventEmitter {
  constructor(config = {}) {
    super();
    
    this.config = {
      nodeCount: config.nodeCount || 20,
      updateInterval: config.updateInterval || 30000, // Changed from 5000 to 15000ms (update every 15 seconds)
      baseTrafficLevel: config.baseTrafficLevel || 0.15,
      trafficVariability: config.trafficVariability || 0.35, // Changed from 0.05 to 0.35 (higher variability)
      incidentProbability: config.incidentProbability || 0.00001,
      roadDensity: config.roadDensity || 0.3 // Probability of connection between nodes
    };
    
    this.vertices = [];
    this.edges = [];
    this.incidents = [];
    this.updateTimer = null;
    this.startTime = Date.now();
    this.updateCount = 0;
    
    this.initializeGraph();
  }
  
  initializeGraph() {
    // Create vertices (intersections/nodes)
    this.vertices = [];
    for (let i = 0; i < this.config.nodeCount; i++) {
      this.vertices.push({
        id: `node_${i}`,
        name: `Intersection ${i + 1}`,
        x: Math.random() * 800 + 50, // For visualization positioning
        y: Math.random() * 500 + 50
      });
    }
    
    // Create edges (roads between nodes)
    this.edges = [];
    const edgeSet = new Set();
    
    // Connect each node to a few random nearby nodes
    for (let i = 0; i < this.vertices.length; i++) {
      const nodeA = this.vertices[i];
      
      // Calculate distances to other nodes
      const distances = this.vertices.map((nodeB, j) => {
        if (i === j) return { index: j, distance: Infinity };
        const dx = nodeA.x - nodeB.x;
        const dy = nodeA.y - nodeB.y;
        return { index: j, distance: Math.sqrt(dx * dx + dy * dy) };
      });
      
      // Sort by distance
      distances.sort((a, b) => a.distance - b.distance);
      
      // Connect to 2-4 nearest neighbors
      const connectCount = 2 + Math.floor(Math.random() * 3);
      for (let k = 0; k < Math.min(connectCount, distances.length); k++) {
        const j = distances[k].index;
        const edgeId = [i, j].sort().join('-');
        
        if (!edgeSet.has(edgeId) && Math.random() < (this.config.roadDensity + 0.5)) {
          edgeSet.add(edgeId);
          
          const distance = distances[k].distance;
          this.edges.push({
            id: `edge_${this.edges.length}`,
            from: nodeA.id,
            to: this.vertices[j].id,
            weight: this.generateRandomTraffic(),
            distance: Math.round(distance / 10), // Scale to reasonable values
            speed: this.calculateSpeed(this.config.baseTrafficLevel),
            vehicleCount: Math.floor(Math.random() * 50),
            isIncident: false,
            incidentSeverity: 0,
            bidirectional: true
          });
        }
      }
    }
    
    // Ensure graph is connected (at least one path between any two nodes)
    this.ensureConnectivity();
  }
  
  ensureConnectivity() {
    // Simple connectivity check - ensure every node has at least one edge
    const connectedNodes = new Set();
    
    for (const edge of this.edges) {
      connectedNodes.add(edge.from);
      connectedNodes.add(edge.to);
    }
    
    // Connect isolated nodes to nearest connected node
    for (const vertex of this.vertices) {
      if (!connectedNodes.has(vertex.id)) {
        // Find nearest connected node
        let nearest = null;
        let minDist = Infinity;
        
        for (const other of this.vertices) {
          if (connectedNodes.has(other.id)) {
            const dx = vertex.x - other.x;
            const dy = vertex.y - other.y;
            const dist = Math.sqrt(dx * dx + dy * dy);
            if (dist < minDist) {
              minDist = dist;
              nearest = other;
            }
          }
        }
        
        if (nearest) {
          this.edges.push({
            id: `edge_${this.edges.length}`,
            from: vertex.id,
            to: nearest.id,
            weight: this.generateRandomTraffic(),
            distance: Math.round(minDist / 10),
            speed: this.calculateSpeed(this.config.baseTrafficLevel),
            vehicleCount: Math.floor(Math.random() * 50),
            isIncident: false,
            incidentSeverity: 0,
            bidirectional: true
          });
          connectedNodes.add(vertex.id);
        }
      }
    }
  }
  
  generateRandomTraffic() {
    const base = this.config.baseTrafficLevel;
    const variability = this.config.trafficVariability;
    // Generate more diverse initial traffic levels with wider spread
    const randomFactor = (Math.random() - 0.5) * variability * 4; // Increased from *2 to *4
    return Math.max(0, Math.min(1, base + randomFactor));
  }
  
  calculateSpeed(trafficLevel) {
    // Speed inversely proportional to traffic level
    // 0 traffic = 60 mph, high traffic = 5 mph
    return Math.round(60 - (trafficLevel * 55));
  }
  
  updateTraffic() {
    this.updateCount++;
    
    // Apply time-based patterns (rush hour simulation)
    const hour = new Date().getHours();
    let timeMultiplier = 1.0;
    
    // Morning rush: 7-9 AM
    if (hour >= 7 && hour <= 9) {
      timeMultiplier = 1.5;
    }
    // Evening rush: 5-7 PM
    else if (hour >= 17 && hour <= 19) {
      timeMultiplier = 1.6;
    }
    // Night time: 11 PM - 5 AM
    else if (hour >= 23 || hour <= 5) {
      timeMultiplier = 0.4;
    }
    
    // Update each edge (road)
    for (const edge of this.edges) {
      // Get connected edges for traffic flow simulation
      const avgNeighborTraffic = this.getAverageNeighborTraffic(edge);
      
      // Slower, more gradual changes
      // High current traffic influence (0.85) for stability and slow transitions
      // Low neighbor influence (0.05) to maintain road independence
      // Small random factor (0.1) for gentle variations
      let randomChange = (Math.random() - 0.5) * 0.2; // -0.1 to +0.1 gentle swing
      let newTraffic = edge.weight * 0.85 + avgNeighborTraffic * 0.05 + randomChange;
      
      // Occasional minor traffic adjustments (5% chance)
      if (Math.random() < 0.05) {
        newTraffic += (Math.random() - 0.5) * 0.15; // Small ±0.075 adjustment
      }
      
      // Apply time multiplier
      newTraffic *= timeMultiplier;
      
      // Clamp between 0 and 1
      newTraffic = Math.max(0, Math.min(1, newTraffic));
      
      edge.weight = newTraffic;
      edge.speed = this.calculateSpeed(newTraffic);
      
      // Update vehicle count based on traffic level with moderate variability
      const baseVehicles = Math.floor(newTraffic * 100);
      edge.vehicleCount = baseVehicles + Math.floor(Math.random() * 10 - 5); // Reduced to ±5 vehicles
      edge.vehicleCount = Math.max(0, edge.vehicleCount);
      
      // Random incident generation on roads
      if (!edge.isIncident && Math.random() < this.config.incidentProbability) {
        this.createIncident(edge.id, Math.random() * 0.5 + 0.3, 30000 + Math.random() * 60000);
      }
    }
    
    // Update incidents
    this.updateIncidents();
    
    // Emit update event
    this.emit('update', this.getCurrentState());
  }
  
  getAverageNeighborTraffic(edge) {
    // Find edges connected to the same nodes
    const connectedEdges = this.edges.filter(e => 
      e.id !== edge.id && (
        e.from === edge.from || e.from === edge.to ||
        e.to === edge.from || e.to === edge.to
      )
    );
    
    if (connectedEdges.length === 0) {
      return this.config.baseTrafficLevel;
    }
    
    const totalTraffic = connectedEdges.reduce((sum, e) => sum + e.weight, 0);
    return totalTraffic / connectedEdges.length;
  }
  
  createIncident(edgeId, severity = 0.5, duration = 60000, incidentType = null) {
    const edge = this.edges.find(e => e.id === edgeId);
    
    if (!edge) {
      throw new Error(`Edge not found: ${edgeId}`);
    }
    
    // Check if edge already has an incident
    if (edge.isIncident) {
      throw new Error(`Edge ${edgeId} already has an active incident`);
    }
    
    // Validate and clamp severity between 0 and 1
    const validatedSeverity = Math.max(0, Math.min(1, severity));
    
    // Validate duration (minimum 10 seconds, maximum 1 hour)
    const validatedDuration = Math.max(10000, Math.min(3600000, duration));
    
    const incident = {
      id: `incident_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
      edgeId: edgeId,
      from: edge.from,
      to: edge.to,
      severity: validatedSeverity,
      startTime: Date.now(),
      endTime: Date.now() + validatedDuration,
      type: incidentType || this.getRandomIncidentType(),
      description: this.getIncidentDescription(incidentType || this.getRandomIncidentType(), validatedSeverity)
    };
    
    this.incidents.push(incident);
    
    edge.isIncident = true;
    edge.incidentSeverity = incident.severity;
    
    // Increase traffic on this road and connected roads
    this.applyIncidentEffect(edgeId, incident.severity);
    
    console.log(`Incident created on ${edgeId} (${edge.from} → ${edge.to}): ${incident.type} with severity ${validatedSeverity.toFixed(2)}`);
    
    // Emit incident created event
    this.emit('incidentCreated', incident);
    
    return incident;
  }
  
  getRandomIncidentType() {
    const types = ['accident', 'construction', 'roadblock', 'weather', 'event'];
    return types[Math.floor(Math.random() * types.length)];
  }
  
  getIncidentDescription(type, severity) {
    const severityText = severity > 0.7 ? 'severe' : severity > 0.4 ? 'moderate' : 'minor';
    const descriptions = {
      'accident': `${severityText.charAt(0).toUpperCase() + severityText.slice(1)} vehicle accident causing delays`,
      'construction': `${severityText.charAt(0).toUpperCase() + severityText.slice(1)} road construction work`,
      'roadblock': `${severityText.charAt(0).toUpperCase() + severityText.slice(1)} road closure or blockage`,
      'weather': `${severityText.charAt(0).toUpperCase() + severityText.slice(1)} weather-related conditions`,
      'event': `${severityText.charAt(0).toUpperCase() + severityText.slice(1)} special event causing congestion`
    };
    return descriptions[type] || `${severityText.charAt(0).toUpperCase() + severityText.slice(1)} traffic incident`;
  }
  
  getAllEdges() {
    return this.edges.map(e => ({
      id: e.id,
      from: e.from,
      to: e.to,
      distance: e.distance,
      currentWeight: e.weight,
      hasIncident: e.isIncident
    }));
  }
  
  getAvailableEdgesForIncident() {
    return this.edges
      .filter(e => !e.isIncident)
      .map(e => ({
        id: e.id,
        from: e.from,
        to: e.to,
        distance: e.distance,
        currentTraffic: e.weight
      }));
  }
  
  clearIncident(incidentId) {
    const incident = this.incidents.find(i => i.id === incidentId);
    
    if (!incident) {
      throw new Error(`Incident not found: ${incidentId}`);
    }
    
    const edge = this.edges.find(e => e.id === incident.edgeId);
    if (edge) {
      edge.isIncident = false;
      edge.incidentSeverity = 0;
    }
    
    this.incidents = this.incidents.filter(i => i.id !== incidentId);
    
    console.log(`Incident ${incidentId} manually cleared from ${incident.edgeId}`);
    
    // Emit incident cleared event
    this.emit('incidentCleared', incident);
    
    return incident;
  }
  
  getActiveIncidents() {
    return this.incidents.map(i => ({
      ...i,
      remainingTime: Math.max(0, i.endTime - Date.now())
    }));
  }
  
  applyIncidentEffect(edgeId, severity) {
    const edge = this.edges.find(e => e.id === edgeId);
    if (!edge) return;
    
    // Higher traffic impact on incident road (0.5 -> 0.8)
    edge.weight = Math.min(1, edge.weight + severity * 0.8);
    
    // Higher impact on connected roads (0.3 -> 0.5)
    const connectedEdges = this.edges.filter(e => 
      e.id !== edge.id && (
        e.from === edge.from || e.from === edge.to ||
        e.to === edge.from || e.to === edge.to
      )
    );
    
    for (const connectedEdge of connectedEdges) {
      connectedEdge.weight = Math.min(1, connectedEdge.weight + severity * 0.5);
    }
  }
  
  updateIncidents() {
    const now = Date.now();
    this.incidents = this.incidents.filter(incident => {
      if (now >= incident.endTime) {
        // Clear incident from edge
        const edge = this.edges.find(e => e.id === incident.edgeId);
        if (edge) {
          edge.isIncident = false;
          edge.incidentSeverity = 0;
        }
        console.log(`Incident cleared on ${incident.edgeId}`);
        return false;
      }
      return true;
    });
  }
  
  getCurrentState() {
    return {
      vertices: this.vertices,
      edges: this.edges.map(e => ({
        id: e.id,
        from: e.from,
        to: e.to,
        weight: e.weight,
        distance: e.distance,
        speed: e.speed,
        vehicleCount: e.vehicleCount,
        isIncident: e.isIncident,
        incidentSeverity: e.incidentSeverity,
        bidirectional: e.bidirectional
      })),
      incidents: this.incidents,
      config: this.config,
      updateCount: this.updateCount,
      timestamp: Date.now()
    };
  }
  
  getEdgeData(edgeId) {
    return this.edges.find(e => e.id === edgeId) || null;
  }
  
  getVertexData(vertexId) {
    return this.vertices.find(v => v.id === vertexId) || null;
  }
  
  getConfig() {
    return { ...this.config };
  }
  
  updateConfig(newConfig) {
    let needsReinit = false;
    
    if (newConfig.nodeCount && newConfig.nodeCount !== this.config.nodeCount) {
      this.config.nodeCount = newConfig.nodeCount;
      needsReinit = true;
    }
    if (newConfig.roadDensity !== undefined) {
      this.config.roadDensity = newConfig.roadDensity;
      needsReinit = true;
    }
    
    if (needsReinit) {
      this.initializeGraph();
    }
    
    if (newConfig.updateInterval) {
      this.config.updateInterval = newConfig.updateInterval;
      if (this.updateTimer) {
        this.stop();
        this.start();
      }
    }
    if (newConfig.baseTrafficLevel !== undefined) {
      this.config.baseTrafficLevel = newConfig.baseTrafficLevel;
    }
    if (newConfig.trafficVariability !== undefined) {
      this.config.trafficVariability = newConfig.trafficVariability;
    }
    if (newConfig.incidentProbability !== undefined) {
      this.config.incidentProbability = newConfig.incidentProbability;
    }
  }
  
  getStats() {
    const totalEdges = this.edges.length;
    const totalVertices = this.vertices.length;
    let totalTraffic = 0;
    let totalVehicles = 0;
    let avgSpeed = 0;
    
    for (const edge of this.edges) {
      totalTraffic += edge.weight;
      totalVehicles += edge.vehicleCount;
      avgSpeed += edge.speed;
    }
    
    return {
      totalVertices,
      totalEdges,
      averageTrafficLevel: (totalTraffic / totalEdges).toFixed(3),
      totalVehicles,
      averageSpeed: Math.round(avgSpeed / totalEdges),
      activeIncidents: this.incidents.length,
      updateCount: this.updateCount,
      uptime: Date.now() - this.startTime
    };
  }
  
  start() {
    if (this.updateTimer) {
      console.log('Traffic generator already running');
      return;
    }
    
    console.log('Starting traffic generator...');
    this.updateTimer = setInterval(() => {
      this.updateTraffic();
    }, this.config.updateInterval);
  }
  
  stop() {
    if (this.updateTimer) {
      clearInterval(this.updateTimer);
      this.updateTimer = null;
      console.log('Traffic generator stopped');
    }
  }
}

module.exports = TrafficGenerator;
